include/matx/core/make_sparse_tensor.h

```diff
@@ -0,0 +1,139 @@
+                   size_t c0sz, CRD *c1, size_t c1sz, POS *p0, size_t p0sz,
+                   POS *p1, size_t p1sz, bool owning) {
+  // Values.
+  raw_pointer_buffer<VAL, matx_allocator<VAL>> bufv{val, vsz, owning};
+  basic_storage<decltype(bufv)> sv{std::move(bufv)};
+  // Coordinates.
+  raw_pointer_buffer<CRD, matx_allocator<CRD>> bufc0{c0, c0sz, owning};
+  raw_pointer_buffer<CRD, matx_allocator<CRD>> bufc1{c1, c1sz, owning};
+  basic_storage<decltype(bufc0)> sc0{std::move(bufc0)};
+  basic_storage<decltype(bufc1)> sc1{std::move(bufc1)};
+  // Positions.
+  raw_pointer_buffer<POS, matx_allocator<POS>> bufp0{p0, p0sz, owning};
+  raw_pointer_buffer<POS, matx_allocator<POS>> bufp1{p1, p1sz, owning};
+  basic_storage<decltype(bufp0)> sp0{std::move(bufp0)};
+  basic_storage<decltype(bufp1)> sp1{std::move(bufp1)};
+  // Sparse tensor in templated format.
+  return sparse_tensor_t<VAL, CRD, POS, TF, decltype(sv), decltype(sc0),
+                         decltype(sp0)>(shape, std::move(sv),
+                                        {std::move(sc0), std::move(sc1)},
+                                        {std::move(sp0), std::move(sp1)});
+}
+
+// Constructs a sparse matrix in COO format directly from the values and
+// coordinates arrays. The entries should be sorted by row, then column.
+// Duplicate entries should not occur. Explicit zeros may be stored.
+template <typename VAL, typename CRD>
+auto make_coo(VAL *val, CRD *row, CRD *col, const index_t (&shape)[2],
+              size_t nse, bool owning = false) {
+  // No positions, although some forms use [0,nse] in first.
+  const size_t vsz = nse * sizeof(VAL);
+  const size_t csz = nse * sizeof(CRD);
+  return make_sparse_matrix<VAL, CRD, int, COO>(
+      shape, val, vsz, row, csz, col, csz, nullptr, 0, nullptr, 0, owning);
+}
+
+// As above for COO, operating directly on MatX 1-dim tensors for the buffers.
+template <typename ValTensor, typename CrdTensor>
+auto make_coo(ValTensor &val, CrdTensor &row, CrdTensor &col,
+              const index_t (&shape)[2], bool owning = false) {
+  const size_t nse = val.Size(0);
+  return make_coo(val.Data(), row.Data(), col.Data(), shape, nse, owning);
+}
+
+// Constructs a sparse matrix in CSR format directly from the values, the row
+// positions, and column coordinates arrays. The entries should be sorted by
+// row, then column. Explicit zeros may be stored.
+template <typename VAL, typename CRD, typename POS>
+auto make_csr(VAL *val, POS *rowp, CRD *col, const index_t (&shape)[2],
+              size_t nse, bool owning = false) {
+  const size_t vsz = nse * sizeof(VAL);
+  const size_t csz = nse * sizeof(CRD);
+  const size_t psz = (shape[0] + 1) * sizeof(POS);
+  return make_sparse_matrix<VAL, CRD, POS, CSR>(
+      shape, val, vsz, nullptr, 0, col, csz, nullptr, 0, rowp, psz, owning);
+}
+
+// As above for CSR, operating directly on MatX 1-dim tensors for the buffers.
+template <typename ValTensor, typename PosTensor, typename CrdTensor>
```

Comment:
How does this one ensure that the data stays there and doesn't go out of scope if it uses `Data()` directly?

Reply:
That clearly indicates a lack of understanding from my part. How can I ensure that ownership and ref counts are properly dealt with when using MatX vectors to provide the 1-dim buffer contents?

Reply:
If you copy a storage buffer the reference count is handled properly. But if you just access .Data() directly from the tensor there is no guarantee that the tensor will stay in scope when it's used. If you want to keep the storage types you can simply copy them via the make_csr parameters instead of raw pointers. This would increment the reference count inside of the storage class (if there is one)

Reply:
Okay, let me redesign the make sparse tensor methods a bit then to ensure that is the case